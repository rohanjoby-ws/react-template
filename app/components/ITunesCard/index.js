/**
 *
 * ITunesCard
 *
 */

import React, { useRef } from 'react';
import PropTypes from 'prop-types';
import styled from 'styled-components';
import 'antd/dist/antd.css';
import { Card, Popover } from 'antd';
import { PlayCircleOutlined, PauseCircleOutlined } from '@ant-design/icons';
import T from '@components/T';
import { colors } from '@app/themes/index';

const CustomCard = styled(Card)`
  && {
    margin: 1.25em 0;
    box-shadow: ${colors.boxShadow};
  }
`;
const Wrapper = styled.div`
  display: flex;
  alig-items: center;
  justify-content: center;
  margin-bottom: 0.75rem;
  gap: 1em;
`;
const CustomButton = styled.button`
  displaty: flex;
  align-items: center;
  justify-content: center;
  background-color: ${colors.buttonBg};
  border: none;
  border-radius: 0.2em;
  cursor: pointer;
`;
const CustomImg = styled.img`
  display: block;
  width: 45%;
  margin: 0.75rem auto;
`;

export function ITunesCard({
  artistName,
  trackName,
  collectionName,
  trackPrice,
  primaryGenreName,
  releaseDate,
  previewUrl,
  artworkUrl100,
  onActionClick
}) {
  const audioRef = useRef(null);

  const playerAction = {
    play: 'PLAY',
    pause: 'PAUSE'
  };
  const handleTrackPlay = (action) => {
    if (action === playerAction.play) {
      audioRef.current.play();
      onActionClick(audioRef);
    }
    if (action === playerAction.pause) {
      audioRef.current.pause();
    }
  };

  const text = <T id="details" type="heading" marginBottom={5} />;
  const content = (
    <>
      <T id="track-price" values={{ price: trackPrice }} />
      <T id="genre-name" values={{ name: primaryGenreName }} />
      <T id="release" values={{ date: releaseDate }} />
    </>
  );
  return (
    <CustomCard data-testid="i-tunes-card">
      <T data-testid="collectionName" id="collection-name" values={{ name: collectionName }} />
      <Popover placement="topRight" title={text} content={content}>
        <CustomImg src={artworkUrl100} alt="artwork" />
      </Popover>
      <Wrapper>
        <CustomButton data-testid="playButton" onClick={() => handleTrackPlay(playerAction.play)}>
          <PlayCircleOutlined style={{ fontSize: '20px' }} />
        </CustomButton>
        <CustomButton data-testid="pauseButton" onClick={() => handleTrackPlay(playerAction.pause)}>
          <PauseCircleOutlined style={{ fontSize: '20px' }} />
        </CustomButton>
      </Wrapper>
      <audio data-testid="audioElement" ref={audioRef} src={previewUrl} onClick={handleTrackPlay} />
      <T data-testid="trackName" id="track-name" values={{ name: trackName }} type="subheading" marginBottom={5} />
      <T data-testid="artistName" id="artist-name" values={{ name: artistName }} />
    </CustomCard>
  );
}

ITunesCard.propTypes = {
  artistName: PropTypes.string,
  trackName: PropTypes.string,
  collectionName: PropTypes.string,
  trackPrice: PropTypes.number,
  primaryGenreName: PropTypes.string,
  releaseDate: PropTypes.string,
  previewUrl: PropTypes.string,
  artworkUrl100: PropTypes.string,
  onActionClick: PropTypes.func
};
ITunesCard.defaultProps = {
  artistName: 'Mark Ronson',
  trackName: 'Uptown Funk (feat. Bruno Mars)',
  collectionName: 'Uptown Special',
  trackPrice: 128,
  primaryGenreName: 'Pop',
  releaseDate: '13/01/2015',
  previewUrl:
    'https://audio-ssl.itunes.apple.com/itunes-assets/AudioPreview125/v4/b9/0a/00/b90a0076-6ac1-0b1d-3693-ff106a79d906/mzaf_10894270267302036162.plus.aac.p.m4a',
  artworkUrl100:
    'https://is2-ssl.mzstatic.com/image/thumb/Music5/v4/22/a0/3e/22a03e12-0c97-b5fc-d693-c02186f6c79d/source/100x100bb.jpg'
};

export default ITunesCard;
