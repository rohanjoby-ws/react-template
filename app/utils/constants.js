export const ACTIONS = {
  PLAY: 'PLAY',
  PAUSE: 'PAUSE'
};
