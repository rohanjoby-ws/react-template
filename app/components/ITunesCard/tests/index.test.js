/**
 *
 * Tests for ITunesCard
 *
 */

import React from 'react';
import { fireEvent } from '@testing-library/dom';
import { renderWithIntl } from '@utils/testUtils';
import { ACTIONS } from '@utils/constants';
import ITunesCard from '../index';
import { mockData } from './mockData';

describe('<ITunesCard />', () => {
  let track, onActionClick;

  beforeEach(() => {
    jest.useFakeTimers();
    track = mockData;
    onActionClick = jest.fn();
  });
  afterEach(() => {
    jest.restoreAllMocks();
  });

  it('should render and match the snapshot', () => {
    const { baseElement } = renderWithIntl(<ITunesCard track={track} onActionClick={onActionClick} />);
    expect(baseElement).toMatchSnapshot();
  });

  it('should contain 1 ITunesCard component', () => {
    const { getAllByTestId } = renderWithIntl(<ITunesCard track={track} onActionClick={onActionClick} />);
    expect(getAllByTestId('itunes-card').length).toBe(1);
  });

  it('should render the song details inside the card', () => {
    const { getByTestId } = renderWithIntl(<ITunesCard track={track} onActionClick={onActionClick} />);
    expect(getByTestId('collectionName')).toHaveTextContent(track.collectionName);
    expect(getByTestId('trackName')).toHaveTextContent(track.trackName);
    expect(getByTestId('artistName')).toHaveTextContent(track.artistName);
  });

  it('should call onAction click on play with audioRef and play', async () => {
    const actionClickHandler = jest.fn();

    const duration = 10.0;
    const currentTime = 1.0;
    const percent = (currentTime / duration) * 100;
    const { getByTestId } = renderWithIntl(<ITunesCard track={track} onActionClick={actionClickHandler} />);

    fireEvent.click(getByTestId('play-button'));

    const audioRef = getByTestId('audio-element');
    audioRef.currentTime = currentTime;
    audioRef.duration = duration;
    expect(actionClickHandler).toBeCalledWith(
      expect.objectContaining({
        current: audioRef
      }),
      ACTIONS.PLAY
    );
    jest.runOnlyPendingTimers();
    expect(getByTestId('progress-bar').querySelector('.ant-progress-bg')).toHaveStyle(`width: ${percent}%`);
  });

  it('should call onAction click on pause with audioRef and pause', (done) => {
    const actionClickHandler = jest.fn();
    const { getByTestId } = renderWithIntl(<ITunesCard track={track} onActionClick={actionClickHandler} />);
    const audioRef = getByTestId('audio-element');
    audioRef.paused = false;
    fireEvent.click(getByTestId('pause-button'));
    expect(actionClickHandler).toBeCalledWith(
      expect.objectContaining({
        current: audioRef
      }),
      ACTIONS.PAUSE
    );
    done();
  });
});
