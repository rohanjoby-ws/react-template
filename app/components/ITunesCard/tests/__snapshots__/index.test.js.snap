// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<ITunesCard /> should render and match the snapshot 1`] = `
<body>
  <div>
    <div
      class="ant-card ant-card-bordered ITunesCard__CustomCard-plw8mc-0 evgQKq"
      data-testid="i-tunes-card"
    >
      <div
        class="ant-card-body"
      >
        <p
          class="T__StyledText-gjlic1-0 kYcyRq"
          data-testid="collectionName"
        >
          Playing from: Uptown Special
        </p>
        <img
          alt="artwork"
          class="ITunesCard__CustomImg-plw8mc-3 kFrJkg"
          src="album-art-url"
        />
        <div
          class="ITunesCard__Wrapper-plw8mc-1 kevQCZ"
        >
          <button
            class="ITunesCard__CustomButton-plw8mc-2 dydca-D"
            data-testid="playButton"
          >
            <span
              aria-label="play-circle"
              class="anticon anticon-play-circle"
              role="img"
              style="font-size: 20px;"
            >
              <svg
                aria-hidden="true"
                data-icon="play-circle"
                fill="currentColor"
                focusable="false"
                height="1em"
                viewBox="64 64 896 896"
                width="1em"
              >
                <path
                  d="M512 64C264.6 64 64 264.6 64 512s200.6 448 448 448 448-200.6 448-448S759.4 64 512 64zm0 820c-205.4 0-372-166.6-372-372s166.6-372 372-372 372 166.6 372 372-166.6 372-372 372z"
                />
                <path
                  d="M719.4 499.1l-296.1-215A15.9 15.9 0 00398 297v430c0 13.1 14.8 20.5 25.3 12.9l296.1-215a15.9 15.9 0 000-25.8zm-257.6 134V390.9L628.5 512 461.8 633.1z"
                />
              </svg>
            </span>
          </button>
          <button
            class="ITunesCard__CustomButton-plw8mc-2 dydca-D"
            data-testid="pauseButton"
          >
            <span
              aria-label="pause-circle"
              class="anticon anticon-pause-circle"
              role="img"
              style="font-size: 20px;"
            >
              <svg
                aria-hidden="true"
                data-icon="pause-circle"
                fill="currentColor"
                focusable="false"
                height="1em"
                viewBox="64 64 896 896"
                width="1em"
              >
                <path
                  d="M512 64C264.6 64 64 264.6 64 512s200.6 448 448 448 448-200.6 448-448S759.4 64 512 64zm0 820c-205.4 0-372-166.6-372-372s166.6-372 372-372 372 166.6 372 372-166.6 372-372 372zm-88-532h-48c-4.4 0-8 3.6-8 8v304c0 4.4 3.6 8 8 8h48c4.4 0 8-3.6 8-8V360c0-4.4-3.6-8-8-8zm224 0h-48c-4.4 0-8 3.6-8 8v304c0 4.4 3.6 8 8 8h48c4.4 0 8-3.6 8-8V360c0-4.4-3.6-8-8-8z"
                />
              </svg>
            </span>
          </button>
        </div>
        <audio
          data-testid="audioElement"
          src="preview-url"
        />
        <p
          class="T__StyledText-gjlic1-0 liNLPV"
          data-testid="trackName"
        >
          Uptown Funk
        </p>
        <p
          class="T__StyledText-gjlic1-0 kYcyRq"
          data-testid="artistName"
        >
          Mark Ronson
        </p>
      </div>
    </div>
  </div>
</body>
`;
